Ensure only valid subdomain names can be used.

xitij2000 opened the pull request:
Currently there is very little validation of subdomain names while registering a beta instance. This means that:
- It is possible to pick subdomain names that are simply invalid (for instance they include an underscore). 
- It is possible to override someone else's subdomain by using different the same name with a different case. 
- It is possible to use dots in the subdomain names, which can have unexpected effects if one user picks a subdomain under another user. 
- It is possible to pick a subdomain name that is longer than the allowed 63 characters.

This pull request includes changes that ensure that it is no longer possible to register domain names that cause the above issues. 

**Sandbox URL**: TBD - sandbox is being provisioned.

**Merge deadline**: None

**Testing instructions**:

1. Set up Ocim using the patches in this pull request. 
2. Visit /registration/ 
3. Try to register an instance that is too long, uses capital letters, a dot or other invalid characters like an underscore. 
4. The registration should fail and the registration form should show the appropriate error message.

**Author notes and concerns**:

Existing instances that don't pass the above validation will need manual intervention. 

**Reviewers**
- [ ] @pomegranited

pomegranited commented on registration/migrations/0006_auto_20171225_0938.py:
```diff
@@ -0,0 +1,77 @@
+import registration.models
+
+
+def migrate_to_lowercase_domains(apps, schema_editor):
```
Could you please split this migration into two parts, one which adds the subdomain validation to the model, and a separate data migration for fixing the existing internal domains?

Then, if you submit the data change as a separate PR, then we can merge the schema change and deploy it, and at least stop any new invalid domains from being created until we can sort out the manual changes required.

pomegranited commented on registration/models.py:
```diff
@@ -79,7 +79,7 @@ class BetaTestApplication(ValidateModelMixin, TimeStampedModel):
     user = models.OneToOneField(settings.AUTH_USER_MODEL, on_delete=models.CASCADE)
     subdomain = models.CharField(
-        max_length=255,
+        max_length=63,
```
Why 63?  (I haven't read the RFC).

Have you checked to see if any of the existing instances' subdomains exceed 63 characters?

Since this change will actually cause the database field to be truncated for all instances (be they archived or not), it's probably best to avoid changing the actual `max_length` and instead just add another validator.

xitij2000 replied:
I did do a check of all the instances. I exported all instances using the API and used a Python script to validate the maximum length and the regex. No subdomains were too long, but a couple failed validation. Those were attached to the ticket. 

I will change this to a validator though, as that will also give us better control over the error message when validation fails.

pomegranited commented on registration/models.py:
```diff
@@ -91,8 +91,9 @@ class BetaTestApplication(ValidateModelMixin, TimeStampedModel):
             validators.RegexValidator(
-                r'^[\w.-]+$',
-                "Please include only letters, numbers, '_', '-' and '.'",
+                r'^[a-zA-Z0-9]([a-zA-Z0-9\-]+[a-zA-Z0-9])?$',
```
We don't need to allow uppercase letters here, if we're just going to lowercase them later anyway.

UmanShahzad replied:
I suppose this is a UX thing: do we want to tell the user they have to lowercase their chars? (This also means the error message has to change).

pomegranited replied:
Yep, I think it's ok to tell the user to lowercase their characters.

We also don't want to have to deal with uniqueness checks that fail *after* we lowercase their domains.

xitij2000 replied:
Perhaps we can accept upper case letters but force the domain name to lower case via JavaScript on blur?

UmanShahzad replied:
That would work but is extra work for not that much gain I think :P

pomegranited commented on registration/tests/test_views.py:
```diff
@@ -224,14 +224,37 @@ def test_valid_application(self):
+        '_hogwarts', 'hog_warts', 'hogwarts_',
+        'hog.warts',
+    )
+    def test_invalid_subdomain(self, bad_subdomain):
```
Please add tests for uppercase subdomains too.

pomegranited commented on registration/tests/test_views.py:
```diff
@@ -224,14 +224,37 @@ def test_valid_application(self):
-            'subdomain': ["Please include only letters, numbers, '_', '-' "
-                          "and '.'"],
+            'subdomain': ['Ensure this value has at most 63 characters (it has 66).'],
         })
```
Please also add a tests that illustrates what happens when uniqueness is violated (including case insensitivity).

xitij2000 replied:
The ``test_existing_subdomain`` already tests for violating uniqueness, and I just realised that with the regex change, it won't be possible to add a subdomain in upper case since it'll already be rejected for violating the regex.

xitij2000 commented:
Sorry about the WIP. I forgot to remove that once I knew I was done. 
I have rebased and pushed code that addresses your comments other than a test for checking case-insensitive uniqueness, which (I think) won't be an issue anymore since subdomains are forced to be lower-case.

pomegranited commented on registration/migrations/0006_auto_20180105_1631.py:
```diff
@@ -0,0 +1,50 @@
+                    ),
+                    django.core.validators.RegexValidator(
+                        '^[a-z0-9]([a-z0-9\\-]+[a-z0-9])?$',
+                        'Please include only lower-case letters, numbers, '
```
nit: since your regex handles the "minimum 3 characters" requirement, I don't ever see the `'The subdomain name must at least have 3 characters.'` error.

I think it's best to keep the `MinLengthValidator` as-is, just in case someone comes along later and changes our regex.  But maybe just update the message to `"Please include at least 3 lower-case..."`

pomegranited commented:
Closed in favour of https://github.com/open-craft/opencraft/pull/263.
